src/bookshelf/src/bksw_sec.erl

```diff
@@ -35,109 +37,283 @@ is_authorized(Req0, #context{} = Context) ->
-    ContentType = proplists:get_value('Content-Type', Headers, ""),
-    Date = proplists:get_value('Date', Headers, ""),
-    %% get_object_and_bucket decodes the bucket, but the request will have been signed with
-    %% the encoded bucket.
-    {ok, Bucket0, Resource} = bksw_util:get_object_and_bucket(Req0),
-    Bucket = bksw_io_names:encode(Bucket0),
-    AccessKey = bksw_conf:access_key_id(Context),
-    SecretKey = bksw_conf:secret_access_key(Context),
-
-    {StringToSign, RawCheckedAuth} =
-        mini_s3:make_authorization(AccessKey, SecretKey,
-                                   erlang:list_to_existing_atom(Method),
-                                   bksw_util:to_string(ContentMD5),
-                                   bksw_util:to_string(ContentType),
-                                   bksw_util:to_string(Date),
-                                   AmzHeaders, bksw_util:to_string(Bucket),
-                                   "/" ++ bksw_util:to_string(Resource),
-                                   ""),
-    CheckedAuth = erlang:iolist_to_binary(RawCheckedAuth),
-    [AccessKeyId, Signature] = split_authorization(IncomingAuth),
-    case erlang:iolist_to_binary(IncomingAuth) of
-        CheckedAuth ->
-            {true, Req0, Context};
+% https://docs.aws.amazon.com/AmazonS3/latest/API/sig-v4-header-based-auth.html
+%-spec check_signed_headers_authhead(proplist(), proplist()) -> boolean(). % for erlang20+
+-spec check_signed_headers_authhead(SignedHeaders::[tuple()], Headers::[tuple()]) -> boolean().
+check_signed_headers_authhead(SignedHeaders, Headers) ->
+%    check_signed_headers_common(SignedHeaders, Headers) andalso
+%
+%    % x-amz-content-sha256 header is required
+%    proplists:is_defined("x-amz-content-sha256", SignedHeaders) andalso
+%
+%    % if content-type header is present in request, it is required
+%    case proplists:is_defined("content-type", Headers) of
+%        true ->
+%            proplists:is_defined("content-type", SignedHeaders);
+%        _ ->
+%            true
+%    end.
+    check_signed_headers_common(SignedHeaders, Headers).
+
+% required signed headers common to both authorization header verification
+% and presigned url verification.
+% https://docs.amazonaws.cn/en_us/AmazonS3/latest/API/sigv4-query-string-auth.html
+%-spec check_signed_headers_common(proplist(), proplist()) -> boolean(). % for erlang20+
+-spec check_signed_headers_common(SignedHeaders::[tuple()], Headers::[tuple()]) -> boolean().
+check_signed_headers_common(SignedHeaders, Headers) ->
+%    % host header is required
+%    proplists:is_defined("host", SignedHeaders) andalso
+%
+%    % any x-amz-* headers present in request are required
+%    [] == [Key || {Key, _} <- Headers, is_amz(Key), not proplists:is_defined(Key, SignedHeaders)].
+
+% any x-amz-* headers present in request are required
+[] == [Key || {Key, _} <- Headers, is_amz(Key), not proplists:is_defined(Key, SignedHeaders)].
```

Comment:
Odd spacing here

Reply:
Do you prefer breaking it into more than one line?